### ISSUE TITLE
Using Kubernetes secret for alpha configuration provider client secret

### ISSUE THREAD

**Issue by albert0815** [oauth2-proxy/manifests#104]
Hi,

When using alpha configuration it seems that using the "existingSecret" is not possible any more. As far as I understand we need to provide the secret of the client in plain text in the provider definition in the alpha configuration. Is this true?

Currently my configuration looks like similar to this:
```
alphaConfig:
          enabled: true
            providers:
              - id: Keycloak
                clientID: oauth2-gateway
                clientSecret: xxxxx
                provider: keycloak-oidc
                name: Keycloak
                oidcConfig:
                  issuerURL: https://keycloak-url
```

I want to know whether it would be possible to use something like this:
```
alphaConfig:
          enabled: true
            providers:
              - id: Keycloak
                existingSecret: keycloak-secret # <- is this possible? This would refer to a k8s secret similar to config.existingsecret.
                provider: keycloak-oidc
                name: Keycloak
                oidcConfig:
                  issuerURL: https://keycloak-url
```

If not, is this something which could be implemented in the chart?

Thanks
Dirk

**Comment by pierluigilenoci** [comment]
I'm sorry but the chart currently doesn't work like that.
Ref: https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/templates/configmap-alpha.yaml#L14-L32

**Comment by albert0815** [comment]
Thank you for your response and the reference. As far as I understand it might actually not be the chart structure itself which is making this impossible. Since alpha configuration is provided as a file into the oauth2-proxy this could only work if oauth2-proxy would allow resolving data in the configuration file itself from environment variables.

If we could pass in something like this into oauth2-proxy:

```
- provider: oidc
  clientSecret: ${OIDC_SECRET_ENVIRONMENT_VARIABLE}
  clientID: oauth2-proxy
  oidcConfig:
    oidcIssuerURL: http://dex.localhost:4190/dex
```

Then it should also be possible to configure it based on the helm chart. Do you think this is something which could be done?

**Comment by pierluigilenoci** [comment]
@albert0815 for my knowledge in Helm I don't think there is room to do as you want. 
At least without overcomplicating the setup.

A possible solution is to load a configMap external to the chart.
Ref: https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/values.yaml#L52

Or using helm post-rendering 
Ref: https://helm.sh/docs/topics/advanced/#post-rendering

Or passing `configData` with the Helm CLI options.
Ref: https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/values.yaml#L50

@desaintmartin what do you think about this?

**Comment by albert0815** [comment]
Thank you for you quick feedback!

My suggested approach with that sample config was actually that the runtime application oauth2-proxy could resolve the enviroment variable after reading the configuration (e.g. by using viper to read the config file with AutomaticEnv enabled).

Regarding your suggested alternatives:

* Config Map / configData: As far as I understand this does not solve my original problem that I do not want to have the clientSecret available in plain text (we actually use configData and need to place the secret in plain text in the configData).
* Post-Rendering: Yes, I guess this would work. In my opinion the client secret is sensitive information and this should be supported out of the box in oauth2-proxy. Before I start implementing a post process I would like to understand whether the oauth2-proxy team agrees that in a production environment client secret is a sensitive information which should be stored encrypted. I assumes this since there is the build in option for the "not alpha" configuration.

**Comment by pierluigilenoci** [comment]
@albert0815 in my opinion the `secrets` should be injected into the pods via some mechanism (ie CSI driver) directly from vaults/secrets managers.

Unfortunately, to my knowledge, there is no way to do the same with `configMaps`.
So, to date, I don't have the perfect solution to suggest.

@JoelSpeed @NickMeves what do you think about that?

**Comment by JoelSpeed** [comment]
In the longer term we need to replace the secret values with a `SecretSource` which has an explicit option to get the secret data from an env var or file. If you wanted to raise a PR to change this in the code, then I'd be happy to review it, but it would be a breaking change to the alpha config, which I'm ok with at the moment, it's something we intend to do

**Comment by pierluigilenoci** [comment]
So @albert0815 feel free to open a PR to solve this. 🚀

**Comment by aarontavio** [comment]
Hi!

I think, you could solve the problem by using the `clientSecretFile` property of the [alpha configuration](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/alpha-config#provider) to specify a file containing the credentials. This file could be mounted into the container by using a kubernetes secret. Now there are multiple ways to handle kubernetes secrets in a proper way so that you don't need to reveal your credentials in plain text, for instance `SealedSecrets` could be an option.

If you use the oauth2-proxy helm chart to deploy your proxy you can use the properties `extraVolumes` & `extraVolumeMounts` of the values file for this.

**Comment by mkjpryor** [comment]
The easiest solution to this is surely just to use a secret instead of a configmap to store the alpha config? Is there a good reason not to do this?

**Comment by mkjpryor** [comment]
Then you can safely put the client secret in there.

**Comment by pierluigilenoci** [comment]
@albert0815 any update on this?

**Comment by pierluigilenoci** [comment]
Solved with #153
